velero-plugins/imagestream/backup.go

```diff
@@ -38,55 +40,151 @@ type BackupPlugin struct {
+		// in imagestream backup plugin: copy all local images image (reverse order per tag):
+		//                                1) if tag has null tag reference, copy to dest:tag
+		//                                2) if there is a tag reference, copy to dest@sha
+		// in imagestream restore plugin: copy all local images image (reverse order per tag):
+		//                                1) if tag has null tag reference, copy to dest:tag
+		//                                2) if there is a tag reference, copy to dest@sha
+		//                                restore imagestream if no local images copied via dest tag
+		specTag := findSpecTag(im.Spec.Tags, tag.Tag)
+		copyToTag := true
+		if specTag != nil && specTag.From != nil {
+			p.Log.Info(fmt.Sprintf("image tagged: %s, %s", specTag.From.Kind, specTag.From.Name))
+			// we have a tag.
+			copyToTag = false
+		}
+		// Iterate over items in reverse order so most recently tagged is copied last
+		for i := len(tag.Items) - 1; i >= 0; i-- {
+			dockerImageReference := tag.Items[i].DockerImageReference
+			localImage := strings.HasPrefix(dockerImageReference, internalRegistry)
+			//p.Log.Info(fmt.Sprintf("image is local?: %t", localImage))
+			if localImage {
+				localImageCopied = true
+				destTag := "@" + tag.Items[i].Image
+				if copyToTag {
+					localImageCopiedByTag = true
+					destTag = ":" + tag.Tag
+				}
+				// This is where the actual image copy will go once skopeo code is vendored
+				// into plugin.
+				p.Log.Info(fmt.Sprintf("skopeo copy --src-creds=<how to get oc user:token?> --src-tls-verify=false --dest-tls-verify=false docker://%s docker://%s/%s/%s%s", dockerImageReference, migrationRegistry, im.Namespace, im.Name, destTag))
+			}
 		}
 	}
-	metadata.SetAnnotations(annotations)
+	p.Log.Info(fmt.Sprintf("copied at least one local image: %t", localImageCopied))
+	p.Log.Info(fmt.Sprintf("copied at least one local image by tag: %t", localImageCopiedByTag))
+
+	im.Annotations = annotations
+
+	var out map[string]interface{}
+	objrec, _ := json.Marshal(im)
+	json.Unmarshal(objrec, &out)
+	item.SetUnstructuredContent(out)
 
 	return item, nil, nil
 }
 
+func findSpecTag(tags []imagev1API.TagReference, name string) *imagev1API.TagReference {
+	for _, tag := range tags {
+		if tag.Name == name {
+			return &tag
+		}
+	}
+	return nil
+}
+
+func findStatusTag(tags []imagev1API.NamedTagEventList, name string) *imagev1API.NamedTagEventList {
+	for _, tag := range tags {
+		if tag.Tag == name {
+			return &tag
+		}
+	}
+	return nil
+}
+
+func getRegistryInfo(p *BackupPlugin, ocpVersion string) (string, error) {
+	cClient, err := p.coreClient()
+	if err != nil {
+		return "", err
+	}
+	if strings.HasPrefix(ocpVersion, "3.") {
+		registrySvc, err := cClient.Services("default").Get("docker-registry", metav1.GetOptions{})
+		if err != nil {
+			return "", err
+		}
+		internalRegistry := registrySvc.Spec.ClusterIP
+		if registrySvc.Spec.Ports[0].Port != 80 {
```

Comment:
Why are we checking to ensure it's not port 80 here?

Reply:
So I think 1) 80 may be the wrong port anyway and 2) it's probably unnecessary. I was trying to account for the case where the dockerImageReferences in the local registry were missing the (optional) port number, but I haven't actually seen any cases where this is true. Even if you use an exposed route for the registry and push/tag externally using this route path (without a port), internally the :5000 gets added, at least from what I've seen. I'll remove the check here and assume the port will always be shown. This is only relevant for 3.x anyway, as for 4.x we can pull the registry path directly from the configMap.